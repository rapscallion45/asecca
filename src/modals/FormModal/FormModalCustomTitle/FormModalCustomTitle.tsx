import { FC, ReactNode } from 'react';
import { IconButton, Box } from '@mui/material';
import MuiDialogTitle from '@mui/material/DialogTitle';
import CloseIcon from '@mui/icons-material/Close';

/**
 * Form Modal Custom Title Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @typedef IFormModalCustomTitleProps
 * @prop {ReactNode} children - child nodes
 * @prop {any} onClose - modal on close callback
 */
interface IFormModalCustomTitleProps {
  children: ReactNode;
  onClose: () => void;
}

/**
 * Form Modal Custom Title
 *
 * Form modal custom title interface
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @component
 * @param {IFormModalCustomTitleProps} props - component props
 * @returns {FC} - form modal custom title functional component
 */
const FormModalCustomTitle: FC<IFormModalCustomTitleProps> = (props) => {
  const { children, onClose, ...other } = props;
  return (
    <MuiDialogTitle {...other}>
      <Box display="flex" alignItems="center">
        {children}
        <Box display="flex" justifyContent="right" sx={{ flexGrow: 1 }}>
          {onClose && (
            <IconButton aria-label="close" onClick={onClose}>
              <CloseIcon />
            </IconButton>
          )}
        </Box>
      </Box>
    </MuiDialogTitle>
  );
};

export default FormModalCustomTitle;
