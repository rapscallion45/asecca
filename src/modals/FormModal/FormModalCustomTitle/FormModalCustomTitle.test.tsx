import React, { FC } from 'react';
import { screen, render, waitFor, fireEvent } from '@testing-library/react';
import FormModalCustomTitle from './FormModalCustomTitle';

/* mock action function callback */
const mockOnCloseCallback = jest.fn(() => {});

/* test child component */
const TestChild: FC = () => <div>Test child</div>;

/**
 * Form Modal Custom Title Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.23
 */
describe('Form Modal Custom Title', () => {
  it('Renders correctly', async () => {
    /** perform snapshot test */
    const tree = render(
      <FormModalCustomTitle onClose={mockOnCloseCallback}>
        <TestChild />
      </FormModalCustomTitle>
    );
    expect(tree).toMatchSnapshot();
  });

  describe('Functions', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should call onClose callback', async () => {
      /** Arrange */
      /** Act */
      render(
        <FormModalCustomTitle onClose={mockOnCloseCallback}>
          <TestChild />
        </FormModalCustomTitle>
      );

      /** Assert - modal button should be rendered and modal displayed */
      await waitFor(() => {
        expect(screen.getByRole('button')).toBeInTheDocument();
      });
      expect(screen.getByTestId('CloseIcon')).toBeInTheDocument();

      /** Act - click on Cancel button to close modal */
      fireEvent(
        screen.getByTestId('CloseIcon'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );

      /** Assert - ensure modal is closed */
      await waitFor(() => {
        expect(mockOnCloseCallback).toHaveBeenCalledTimes(1);
      });
    });
  });
});
