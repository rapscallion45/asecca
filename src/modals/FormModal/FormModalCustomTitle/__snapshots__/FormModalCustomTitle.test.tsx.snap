// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Modal Custom Title Renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2
        class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <div>
            Test child
          </div>
          <div
            class="MuiBox-root css-17a0awl"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </h2>
    </div>
  </body>,
  "container": <div>
    <h2
      class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <div>
          Test child
        </div>
        <div
          class="MuiBox-root css-17a0awl"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </h2>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
